File: app-python/entitlements_app/services/add_feature_to_plan_service.py
"""Grant a feature allowance to a plan."""

from decimal import Decimal
import re

from fastapi import HTTPException, status

from entitlements_app.requests.add_feature_to_plan_request import AddFeatureToPlanRequest
from entitlements_app.entities import Entitlement
from entitlements_app.ports import EntitlementsRepository


class AddFeatureToPlanService:
    """Service handling Command `AddFeatureToPlan`."""

    def __init__(self, entitlements_repository: EntitlementsRepository) -> None:
        self.entitlements_repository: EntitlementsRepository = entitlements_repository

    def handle(self, request: AddFeatureToPlanRequest) -> Entitlement:
        # spec-driven body for Command AddFeatureToPlan.
        # intent: Grant a feature allowance to a plan.
        # requires: EntitlementsRepository
        # emits events: EntitlementGranted
        # return: Entitlement (entity); construct from request inputs and persist via EntitlementsRepository.saveEntitlement.
        value = Entitlement(tenantId=request.tenantId, entitlementId=request.entitlementId, planId=request.planId, featureKey=request.featureKey, allowance=request.allowance, unit=request.unit, enforcementMode=request.enforcementMode, effectiveAt=request.effectiveAt)
        return self.entitlements_repository.saveEntitlement(value)

File: app/src/main/java/com/example/spring/AddFeatureToPlanService.java
package com.example.spring;

import com.example.spring.error.ConflictException;
import com.example.spring.error.NotFoundException;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Grant a feature allowance to a plan.
 */
@Service
public class AddFeatureToPlanService {

    private final EntitlementsRepository entitlementsRepository;
    private final ApplicationEventPublisher springEventPublisher;

    public AddFeatureToPlanService(EntitlementsRepository entitlementsRepository, ApplicationEventPublisher springEventPublisher) {
        this.entitlementsRepository = entitlementsRepository;
        this.springEventPublisher = springEventPublisher;
    }

    @Transactional
    public Entitlement handle(AddFeatureToPlanRequest request) {
        // === GEN_REGION_START: service-body ===
        // spec-driven body for AddFeatureToPlan.
        // intent: Grant a feature allowance to a plan.
        // requires: EntitlementsRepository
        // emits events: EntitlementGranted
        // return: Entitlement (entity); spec-driven body constructs from request inputs + persists via EntitlementsRepository.saveEntitlement.
        Entitlement value = new Entitlement(request.tenantId(), request.entitlementId(), request.planId(), request.featureKey(), request.allowance(), request.unit(), request.enforcementMode(), request.effectiveAt());
        Entitlement saved = this.entitlementsRepository.saveEntitlement(value);
        this.springEventPublisher.publishEvent(new EntitlementGranted(saved.allowance(), saved.effectiveAt(), saved.enforcementMode(), saved.entitlementId(), saved.featureKey(), saved.planId(), saved.tenantId(), saved.unit()));
        return saved;
        // === GEN_REGION_END: service-body ===
    }
}